### PR TITLE
Windows packaging build with Gitlab CI

### PR DESCRIPTION
**Pull request by maximedenes**
This PR makes it possible to use a specific runner on Inria CloudStack to automatically generate signed Windows packages (for protected branches, PRs will generate unsigned packages once we set up the Gitlab fork for PR mirroring).

### CODE REVIEW

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: Might be better in stage: test to avoid holding those jobs up.

**Review by ejgallego — APPROVED** [review]
Looks good to me.

**Comment by ejgallego** [comment]
Should this be allowed to fail as to be more resilient wrt INRIA Cloudstack going down?

**Comment by ejgallego** [comment]
Let me know when you want this in.

**Comment by MSoegtropIMC** [comment]
@maximedenes : If I understand this right, the secrets are only available on special INRIA owned machines and running CI on these machines requires special privileges. Can you describe what these privileges are and how you ensure that only INRIA folks allowed to sign binaries in the name of INRIA obtain such privileges?

In general I have my doubts that automating signing is worth the effort. Thinking about how to make this secure is a lot of effort - it is just too easy to miss something. As Edgar Allen Poe wrote 175 years back:

> Few persons can be made to believe that it is not quite an easy thing to invent a method of secret writing which shall baffle investigation. Yet it may be roundly asserted that human ingenuity cannot concoct a cipher which human ingenuity cannot resolve..."

The same applies to all security. I think it makes sense to have a human in the loop for signing.

Technically the  changes are fine, but I have the security concerns stated above and these code changes are by far not sufficient to judge the security of this.

Best regards,

Michael

**Review by SkySkimmer — DISMISSED** [review]
Like with #6962 I am concerned that the build will fail for anyone who doesn't have a tagged runner registered. eg https://gitlab.com/SkySkimmer/coq/pipelines/21065247 currently stalled for lack of runner, I'll kill it at some point if it doesn't die by itself.

I have considered the following solutions:
- we just make it easy to request a runner (simple but makes it harder to use gitlab CI for newcomers, I don't like this solution)
- we use `if: $WINDOWS == "enabled"` and people with access to a windows runner put it in their project variables. This works for this PR but not for #6962 as that one tags jobs that the rest depend upon (so not running them breaks everything). cf https://docs.gitlab.com/ee/ci/variables/README.html#variables-expressions https://gitlab.com/help/ci/variables/README#secret-variables
- (can be seen as an extension of the above) we have a `.gitlab-tagged.yml` using `include: .gitlab-ci.yml` and containing tagged copies of the normal jobs and the windows job, and people with tagged runners can set their project to use gitlab-tagged. cf https://gitlab.com/help/user/project/pipelines/settings#custom-ci-config-path

I think 3 is overkill for this PR so I request you implement 2.

**Comment by SkySkimmer** [comment]
Also I set the WIN_CERTIFICATE_PASSWORD variable to protected on gitlab/coq in case a wild `printenv` appears.

**Comment by SkySkimmer** [comment]
Some interesting stuff in this log https://gitlab.com/coq/coq/-/jobs/64917989
- Coq's make install complains about not having native compute stuff to install
- mingw stuff is downloaded again after bignums is finished (why?)

File: .gitlab-ci.yml
Comment (SkySkimmer) [reply .gitlab-ci.yml]: The artifact upload seems to take a while so definitely put the job in the most parallel stage.

**Comment by MSoegtropIMC** [comment]
@SkySkimmer : the warnings about native compute a normal - there is no native compute on Windows as yet (I am working on it, it currently would need a fully working bash and this mean a fully working Cygwin)

The download of mingw stuff are the sources. These are part of the open source package created together with the installer. The point of this is that GPL stuff requires that you make the sources included in your binaries accessible - redirections are not allowed. So there is a mechanism which takes care to get each and every source file included in the binary build artefacts, and this means sources of all MinGW libraries.

What concerns me more is this line:

Saving to: ‘V8.8+beta1.zip’

I think I requested this to be fixed ... (it is not a reasonable name for a module source package - it should be renamed). I will create an issue to fix it.

**Comment by SkySkimmer** [comment]
Weird error in travis OSX job.

**Comment by SkySkimmer** [comment]
>WARNING: Uploading artifacts to coordinator... failed  id=64917989 responseStatus=504 Gateway Time-out status=504 Gateway Time-out token=TxH_LzFc
> RETRY x2 failed

The windows artifacts are <500MB on appveyor, ~120MB for the regular gitlab artifacts. So it's not orders of magnitude more but it's still a possible failure.
Does the windows runner have good upload speed?
Maybe splitting the artifacts into multiple archives would workaround the timeout (ie just keep the original archives as-downloaded).

It's concerning that artifact upload failure doesn't fail the job. Maybe we should have a second job depending on this one building some arbitrary development (plus it would solve #7171).

>The download of mingw stuff are the sources. These are part of the open source package created together with the installer. The point of this is that GPL stuff requires that you make the sources included in your binaries accessible - redirections are not allowed. So there is a mechanism which takes care to get each and every source file included in the binary build artefacts, and this means sources of all MinGW libraries.

Could we keep the sources on some inria server and avoid repacking them at every job? Especially since it's a likely cause of the failure above.

**Comment by SkySkimmer** [comment]
> https://gitlab.com/SkySkimmer/coq/pipelines/21065247 currently stalled for lack of runner, I'll kill it at some point if it doesn't die by itself.

It died on its own.

**Comment by MSoegtropIMC** [comment]
> Could we keep the sources on some inria server and avoid repacking them at every job? Especially since it's a likely cause of the failure above.

Well it is part of the process of building an installer as much as building the installer itself. The files change frequently (about once a week), so if one separates this from building the installer, it might be hard to track the relation. In the end there is likely more reason to regression test this than to regression test the build of the installer itself.

But what would be great is to have a way to keep copies of all downloaded artifacts on some INRIA server and use this as backup source. I planned to have this on a private Amazon S3 bucket.

**Comment by maximedenes** [comment]
> In general I have my doubts that automating signing is worth the effort.

Well, I've been signing packages for a few years now, and honestly I would prefer to spend this time on development.

> If I understand this right, the secrets are only available on special INRIA owned machines and running CI on these machines requires special privileges. Can you describe what these privileges are and how you ensure that only INRIA folks allowed to sign binaries in the name of INRIA obtain such privileges?

The runners that have the secret are used only on the protected branches of the main gitlab Coq repo. So folks who get the privilege to sign are people who have push access to that gitlab repo, or the GitHub repo (because of the mirroring).

File: .gitlab-ci.yml
Comment (maximedenes) [reply .gitlab-ci.yml]: I also thought I could do a later `package` stage. Especially because it is going to be required on macOS anyway, I think. Does it sound good?

**Comment by maximedenes** [comment]
> Also I set the WIN_CERTIFICATE_PASSWORD variable to protected on gitlab/coq in case a wild printenv appears.

Sure, I had done it but disabled just while this PR was being tested.

**Comment by maximedenes** [comment]
> Coq's make install complains about not having native compute stuff to install
> mingw stuff is downloaded again after bignums is finished (why?)

This is interesting but should be fixed separately.

File: .gitlab-ci.yml
Comment (SkySkimmer) [reply .gitlab-ci.yml]: >I also thought I could do a later package stage. Especially because it is going to be required on macOS anyway, I think. Does it sound good?

It's less parallel, what's the benefit? Also if any of the tests break it will block the package stage.
Why would it be required on apple?

**Comment by Zimmi48** [comment]
> The runners that have the secret are used only on the protected branches of the main gitlab Coq repo.

How do you ensure this? It is because we intend to push branches corresponding to PRs on another GitLab repo, right? We should document this (maybe in this code).

**Comment by maximedenes** [comment]
> How do you ensure this? It is because we intend to push branches corresponding to PRs on another GitLab repo, right? We should document this (maybe in this code).

Even if we push those branches, they wouldn't be protected, would they?

**Comment by MSoegtropIMC** [comment]
> The runners that have the secret are used only on the protected branches of the main gitlab Coq repo. So folks who get the privilege to sign are people who have push access to that gitlab repo, or the GitHub repo (because of the mirroring).

I see. Well you have to decide if this is in accordance with your policies of handling signature keys, including the policies of all institutions involved in issuing the key. In my experience it is already a lot of work to find out what the policies are you need to comply to. Typically if you have a key which is signed by an authority trusted by default by Windows, you have signed something that you take reasonable care that nobody gets access to the key, who is not part of the legal entity to which the key has been issued. If you are lucky, not complying to this policy just leads to revocation of the certificate by the issuer (when they find out). If you are unlucky, there are substantial fines for negligent key leakage. If you considered this already, fine. If not, please check the trust chain of your key and the policies of the parties involved.

sorry for making your life harder!

**Comment by SkySkimmer** [comment]
Sorry it's `only:` https://docs.gitlab.com/ee/ci/yaml/#only-and-except-complex it seems I keep confusing gitlab and travis today.

**Comment by maximedenes** [comment]
> I see. Well you have to decide if this is in accordance with your policies of handling signature keys, including the policies of all institutions involved in issuing the key. In my experience it is already a lot of work to find out what the policies are you need to comply to. Typically if you have a key which is signed by an authority trusted by default by Windows, you have signed something that you take reasonable care that nobody gets access to the key, who is not part of the legal entity to which the key has been issued. If you are lucky, not complying to this policy just leads to revocation of the certificate by the issuer (when they find out). If you are unlucky, there are substantial fines for negligent key leakage. If you considered this already, fine. If not, please check the trust chain of your key and the policies of the parties involved.
> 
> sorry for making your life harder!

I understand your point of view, and I think you are right™. However, I'm afraid we have to be pragmatic: Inria does not have a team responsible for prompt signing, and won't let Coq developers access the key. So we got a key from the Inria foundation. As of today, I'm the only Coq developer part of that entity. My boss (who signed the forms for the key) is ok with they certificate being used as part of an automated process (it is already in use for other software). I don't want to have to go through a manual process relying on two VMs each time we do a release. So either we do things in a way that is not fully right, or we simply don't sign our binaries (I'm also fine with that, but IMHO it would be a regression for our users).

**Comment by maximedenes** [comment]
> negligent key leakage

I hope we can make the argument that mandating core developers to produce signed binaries is not the same as leaking the key in an uncontrolled way.

**Comment by Zimmi48** [comment]
> My boss (who signed the forms for the key) is ok with they certificate being used as part of an automated process (it is already in use for other software).

That's a key point.

**Comment by MSoegtropIMC** [comment]
> I hope we can make the argument that mandating core developers to produce signed binaries is not the same as leaking the key in an uncontrolled way.

I would say this depends on to whom and by whom the key is issued. If the key is issued to say INRIA itself, giving anybody outside of INRIA the possibility to sign using this key could be called negligent key leakage. If the key is issued to say "The Coq Foundation" and not signed with a key owned by INRIA but by an independent authority, it should be fine that everyone who has write rights to the Coq repo can use this key for signing. In case the key is issued to "The Coq Foundation" but signed by a key issued to INRIA, I would say it is a corner case. The key INRIA has might not allow INRIA to issue keys for use outside of their organization - one would have to read the fine print.

But of cause I cannot give you legal counsel. All I say is that matters might be more complicated than you think. At least you should review to whom *exactly* and by whom the key is issued and read the fine print of the contract you signed to get the key. If you have no doubts after this, fine. Otherwise I would discuss this once more with your boss / your legal department / ... or possibly get a key which matches the requirements.

The whole trust system of Windows relies on the idea that if someone messes with your system, at least you can find out whom to sue. If you install software signed by people where it is pretty unknown who they are, it is your own problem. But in my opinion people have a right to reasonably trust software signed by large organizations, and the policies of trusted authorities typically reflect this. 

Once this mechanism is installed and working it is not very likely to be reviewed again. So better get it right.

**Comment by maximedenes** [comment]
In fact, there is an easy solution:
- separate building and signing
- move the runner with secret to my own fork
- make the signing stage require a manual action (Gitlab supports it)

That way, I'll be the one manually producing the signed packages (and having access to the key), but it will require only clicking a button. It should address @MSoegtropIMC 's legal concerns.

**Comment by MSoegtropIMC** [comment]
Indeed this is likely much less work than going through the fine print of the contract.

A good solution!

**Comment by maximedenes** [comment]
I addressed @SkySkimmer 's remarks. The current version no longer relies on secrets for the main repo.

File: .gitlab-ci.yml
Comment (Zimmi48) [inline .gitlab-ci.yml]: IIRC the download and installation of dependencies take a long time in the whole packaging process. In the future, we could end up saving a lot of time by using a personalized Docker image, especially if it is cached on the runners. (Not a request for this PR.)

File: .gitlab-ci.yml
Comment (Zimmi48) [inline .gitlab-ci.yml]: It is for testing only that you only put "64" and you plan to also add "32" later?

File: dev/ci/gitlab.bat
Comment (Zimmi48) [inline dev/ci/gitlab.bat]: You could also make the logs of your GitLab fork private, or have the secrets on a third, private repo, say "gitlab.com/inria-foundation/coq" or "gitlab.com/coq-consortium/coq".

**Review by Zimmi48 — COMMENTED** [review]
I suppose you plan to strip off `dev/ci/appveyor.bat` and `appveyor.yml` later on.

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: It seems the job wasn't created, did you set the variable after pushing or is there a problem?
(also the build job failed and you need to rebase)

Comment (maximedenes) [reply .gitlab-ci.yml]: Yeah I set it after. I should probably relaunch it.

File: .gitlab-ci.yml
Comment (maximedenes) [reply .gitlab-ci.yml]: How does Docker work on Windows?

File: .gitlab-ci.yml
Comment (maximedenes) [reply .gitlab-ci.yml]: Yeah, I forgot to create a second job. Will fix.

**Comment by SkySkimmer** [comment]
You need to rebase so that the linux jobs pass (they block the test stage)

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: I would just call the jobs `windowsxx`

**Comment by MSoegtropIMC** [comment]
One more note: the creation of the open source package is only required for official downloadable installers. So it would make sense to do this only when signing. The problem is that Cygwin changes frequently and it could happen that the packages changed between build and signing - it can even happen that the source packages used for building don't exist any more 2 days later. Cygwin keeps the latest 2 versions and if there are a few updates in 2 days, there will be issues.

This would be solved by using a fixed Cygwin from a repo we control for each version, as I suggested before.

**Comment by maximedenes** [comment]
> One more note: the creation of the open source package is only required for official downloadable installers. So it would make sense to do this only when signing.

Good point, I will make it configurable.

> The problem is that Cygwin changes frequently and it could happen that the packages changed between build and signing

For now, I will do the signing step rebuild the binary. Not a big deal, as long as everything is automated. That will enforce that the opensource archive is in sync with the released binary.

File: .gitlab-ci.yml
Comment (maximedenes) [reply .gitlab-ci.yml]: Done, and rebased.

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: SETUP is reset in the gitlab.bat so not needed here.

Comment (maximedenes) [reply .gitlab-ci.yml]: It's now used in `.gitlab-ci.yml`. I should probably remove it from `gitlab.bat`.

Comment (SkySkimmer) [reply .gitlab-ci.yml]: Why not include the download in the .bat?

Comment (maximedenes) [reply .gitlab-ci.yml]: How?

File: .gitlab-ci.yml
Comment (Zimmi48) [reply .gitlab-ci.yml]: No experience with it but it seems to be supported.

File: .gitlab-ci.yml
Comment (maximedenes) [reply .gitlab-ci.yml]: No, the part that performs the download is PowerShell.

**Comment by SkySkimmer** [comment]
Is it OK to keep the open source files only as long as we keep the binaries (1 week currently)?

File: .gitlab-ci.yml
Comment (SkySkimmer) [reply .gitlab-ci.yml]: Can powershell not be called from bat?

Comment (maximedenes) [reply .gitlab-ci.yml]: I have no idea, I know nothing about Windows dev.

Comment (maximedenes) [reply .gitlab-ci.yml]: But anyway, this download is setting up dependencies, so I'm not shocked by seeing it in `.gitlab-ci.yml`. Isn't it what we do for other builds?

**Comment by Zimmi48** [comment]
Yes, because there are various ways to satisfy the requirements of the GPL but one of them is to propose to download everything at once from the same place.

**Comment by Zimmi48** [comment]
Concretely this means that if a user downloads the binary but not the sources and after one week wants to distribute the binary further, they could get in trouble (not us). If they downloaded both, then no problem. This is all very theoretical. No one is ever going to sue for this.

**Comment by MSoegtropIMC** [comment]
@Zimmi48 : The open source archive is there to fulfill legal requirements - it has not real practical use. I would think that one fulfills the word and spirit of the GPL, if one makes the open source archive only available for official releases and not for intermediate test builds, at least if these official releases are regular. If this is questionable, I could go and ask the GNU lawyers.

Again all this would be solved by making sure that the contents of the open source archive wouldn't change over a release. I suggested to download all this from a Coq specific Repo, which is updated shortly after each release to the latest version. The open source archive can then be created when this repo changes. It doesn't make that much sense to recreate it a few 1000 times on each CI run.

**Comment by SkySkimmer** [comment]
>Exception calling "DownloadFile" with "2" argument(s): "An exception occurred d
uring a WebClient request."
At line:1 char:40
>+ (New-Object Net.WebClient).DownloadFile <<<< ('http://www.cygwin.com/$SETUP',
 $SETUP)
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : DotNetMethodException

**Comment by MSoegtropIMC** [comment]
P.S.: the main benefit of not changing Cygwin any time in the release cycle is that a change from cygwin can't break our CI. Solving the open source archive issue is just a side benefit.

**Comment by Zimmi48** [comment]
> official releases and not for intermediate test builds

As soon as these test builds are available for general download, I'm afraid the requirements kick in.

> Again all this would be solved by making sure that the contents of the open source archive wouldn't change over a release. I suggested to download all this from a Coq specific Repo, which is updated shortly after each release to the latest version. The open source archive can then be created when this repo changes. It doesn't make that much sense to recreate it a few 1000 times on each CI run.

I fully agree with this view and solution, this is also what I had in mind above when I suggested to use a homemade docker image. Unfortunately, I have very little experience with Docker and zero experience with Docker for Windows, so I don't know how doable this proposal here. But even if not done with Docker but rather in a ad-hoc way, I would still support this solution.

**Comment by SkySkimmer** [comment]
@maximedenes while trying to get it to work you may want to run windows in the first stage, so that you avoid wasting time waiting for linux jobs
Maybe even make a new stage before all the others, or delete the other jobs.

**Comment by maximedenes** [comment]
> @maximedenes while trying to get it to work you may want to run windows in the first stage, so that you avoid wasting time waiting for linux jobs
> Maybe even make a new stage before all the others, or delete the other jobs.

That's ok, I'm multitasking.

**Comment by MSoegtropIMC** [comment]
@Zimm48: Using docker with locaI repo is an interesting approach. I have some experience with docker, so please feel free to create a ticket for this and assign it to me.

File: dev/build/windows/makecoq_mingw.sh
Comment (MSoegtropIMC) [inline dev/build/windows/makecoq_mingw.sh]: Can you explain under which circumstances FINDLIBDIR contains backslashes? 
FINDLIBDIR should be in what I call MFMT.
See the docs in MakeCoq_MinGW.bat:

REM WFMT = windows format (C:\..)          Used in this batch file.
REM CFMT = cygwin format (\cygdrive\c\..)  Used for Cygwin PATH varible, which is : separated, so C: doesn't work.
REM MFMT = MinGW format (C:/...)           Used for the build, because \\ requires escaping. Mingw can handle \ and /.

**Comment by SkySkimmer** [comment]
still broken

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: Needs to be $env:foo or %foo% (not sure which).

**Comment by SkySkimmer** [comment]
Fixed the setup download.
It seems we have only the one windows runner, with 2 jobs and 1h15 per job that's going to be quite slow.
The artifact timeout from before scares me too.

File: dev/build/windows/makecoq_mingw.sh
Comment (maximedenes) [reply dev/build/windows/makecoq_mingw.sh]: This path is built by lablgtk calling `ocamlfind`, so we don't have control over the format.

**Comment by SkySkimmer** [comment]
Artifact name looks fixed.
Please move the jobs back to the `test` stage. Also squash.
Do we want both jobs considering how long they take?

**Comment by maximedenes** [comment]
I'll move the job, squash and register 2 runners.

**Comment by Zimmi48** [comment]
Please remove the packaging jobs from AppVeyor at the same time. Given how slow AppVeyor is to build pull requests compared to the rest of our CI, we could use the speedup.

**Comment by maximedenes** [comment]
Squashed, rebased, now with 2 workers, moved to test stage, removed packaging jobs from AppVeyor. Please merge if CI is green.

**Comment by ejgallego** [comment]
Can you update naming to `windows:32bit` and `windows:64bit`?

[we cannot use `ci:foo` for Make reasons tho]

**Comment by maximedenes** [comment]
`32bit` or `32bits`?

**Comment by ejgallego** [comment]
We are not yet fully settled on a consiste naming scheme, opam uses the `32bit` prefix so that's what we have used for now.

**Comment by maximedenes** [comment]
I think this is going to make the build failed because it will produce an invalid file name under windows (because of colon). So I'd rather keep things as they are.

**Comment by Zimmi48** [comment]
In the commit message, you say "a one click operation" on your repo, but it is not obvious to me what the setup for this one click operation is. However, it is rather clear why it will sign only on your repo only: the secrets will be there and not elsewhere.

The removal of the packaging jobs from AppVeyor could have implied a clean-up of the AppVeyor batch script too. But this clean-up can also be done later.

When this is merged, it should probably get a bit of documentation, probably as part of #6057.

**Comment by Zimmi48** [comment]
Did you make sure that the build was also working on your repo when the signing secrets are there?

**Comment by SkySkimmer** [comment]
The runner failed:
>'"git"' is not recognized as an internal or external command,
operable program or batch file.

**Comment by maximedenes** [comment]
Working on it.

**Comment by maximedenes** [comment]
> Did you make sure that the build was also working on your repo when the signing secrets are there?

This PR was tested with and without the secrets, though not on my repo. I plan to setup the "one click" operation as a deployment environment on my repo, making another PR here if needed. This PR, in its current state, does not introduce any regression on that matter, since signing was already done outside of CI.

**Comment by ejgallego** [comment]
> I think this is going to make the build failed because it will produce an invalid file name under windows (because of colon). So I'd rather keep things as they are.

I would recommend not depending on the `CI JOB NAME` as to name the artifacts, but if you want to do that for now that's OK.

**Comment by maximedenes** [comment]
@SkySkimmer I believe this is finally ready.

**Comment by ejgallego** [comment]
It seems that windows runners can only run one concurrent job, pity as this makes scaling a bit more painful.

Anyways, we should really increase the number of windows runners as the current capacity is clearly not enough.

**Pull request merged**
